### DISCUSSION TITLE
Filtering a recursive file copy

### DISCUSSION THREAD

**Discussion by gar1t** [nushell/nushell#10343] [Q&A]
I'm imaging the use of `ls` and `where` to generate a list of files to copy and piping that to a command that copies the files, ala a recursive copy.

Something like this:

    > ls **/* | where size > 1MB | cp /tmp/dest

This seems like a sensible application of Nushell's data passing and command composition vision. Is there an existing approach to this that doesn't involve complex expressions (e.g. the use of `each` or similar)?

**Comment by fdncred** [comment]
Not that I know of, because if you return a list, you have to tell it what to do with each item in the list. We try not to automagically do things because it leads to confusion. So, for your command above, I'd either use `ls` or `glob` and pipe them through each. Something like this:
```nushell
ls **/* | where size > 1MB | each {cp $in.name /tmp/dest}
```
or with glob
```nushell
glob **/* --not [some glob] | each {cp $in /tmp/dest}
```

I think those would work but I haven't tested them.

**Reply by gar1t** [reply]
The issue with those examples is with nested structures.

A new command, e.g. `install` (comparable to `^install`) could do the right thing here (the use of `cp` was just an example in concept).

**Reply by fdncred** [reply]
maybe the new `ucp` (coreutils builtin cp, coming in 0.85) would help more with nested structures. I'm not sure what you want it to do really as i have no clue what `install` is or does.

**Reply by gar1t** [reply]
`install` copies files and will create leading dirs as needed when `-D` is provided. Here's what a file list copy might look like in bash:

    $ find $src -type f | xargs -i install -D {} $dest/{}

The structure under `$dest` will mirror that of `$src`.

This is classic problem for `rsync, which, for the case I raise originally, has file size filter options. But it's not cross platform and it doesn't leverage the flexibility of composition in Nushell.

I don't know what `ucp` is or does :)

I looked for an RFC style doc or PR or discussion thread talking about ucp but didn't turn anything up. Does this work typically happen on the Discord server?

**Reply by fdncred** [reply]
This is the PR https://github.com/nushell/nushell/pull/10097

`ucp` will be released in the next few days with 0.85.0, but it's in the current main now if you want to play around with it.
![image](https://github.com/nushell/nushell/assets/343840/6a62e4c6-066c-4303-932a-90607a941767)

**Comment by gar1t** [comment]
Naively I'd see something like this landing in `cp`.

Currently, this copies a directory structure:

    > cp -r src dest

Hypothetically, this copies a filtered directory structure:

    > ls src/**/*.txt | cp -r dest

`cp` needs to resolve the source root directory for the list, but could assume current dir and use an option otherwise.

Currently, this tries to copy a flat list of files, even with `-r` (it appears that `-r` is ignored when source resolves to more than one file - in this case dest must exist):

    > cp -r src/**/*.txt dest

`cp` currently supports recursive copy and also copies lists of files resolved by a pattern. It seems an appropriate command for consideration of piped input.
 
As an aside, it looks like `coreutils` could have a potentially big impact on Nushell - e.g. lots of `u<cmd>` command variants, creating a parallel shell command set. As this coreutils mirrors the GNU and Nushell uses the `^<cmd>` pattern to escape to system commands, could an overlay be used to apply coreutils-implemented commands into an environment?

The `u` prefix as a general pattern could lead to confusion, e.g. mapping [`install`](https://uutils.github.io/coreutils/book/utils/install.html) becomes `uinstall`.

The cross platform nature of coreutils is a really compelling value prop for Nushell. Personally I'd be delighted to work with a `^<cmd>` syntax even if the CLI is not 100% GNU compatible. (I've gone off topic!)

**Reply by fdncred** [reply]
The `u` prefix is just for testing. It will eventually replace `cp`. We do not plan to maintain 2 `cp` commands. There's currently a bug in coreutils cp command, so we wanted to test it a bit more and give the coreutils team time to fix that bug.
